crates/solidity/testing/sourcify/src/sourcify.rs

```diff
@@ -0,0 +1,323 @@
+use std::fs;
+use std::io::{BufReader, ErrorKind, Read};
+use std::path::{Path, PathBuf};
+
+use anyhow::{bail, Error, Result};
+use reqwest::blocking::Client;
+use semver::Version;
+use serde::Deserialize;
+use slang_solidity::compilation::CompilationUnit;
+use tar::Archive;
+
+use crate::chains::Chain;
+use crate::command::ShardingOptions;
+use crate::compilation_builder::CompilationBuilder;
+use crate::metadata::ContractMetadata;
+
+/// A `Repository` is used to fetch contracts from a certain chain. In addition to providing
+/// an API for fetching contracts, it also manages a directory where fetched contract archives will
+/// be written.
+pub struct Repository {
+    /// The chain that this Repository will fetch from
+    chain: Chain,
+    /// The parent path where fetched `ContractArchives` will be unpacked to.
+    path: PathBuf,
+    /// If `true`, then the files for this repository will not be deleted after the test is complete.
+    should_save: bool,
+}
+
+impl Repository {
+    pub fn new(chain: Chain, should_save: bool) -> Result<Repository> {
+        let root = if should_save {
+            PathBuf::from("target")
+        } else {
+            std::env::temp_dir()
+        };
+
+        let path = root.join(format!("sourcify_{}", chain.id()));
+        create_or_replace_dir(&path)?;
+
+        Ok(Repository {
+            chain,
+            path,
+            should_save,
+        })
+    }
+
+    /// Fetch shard info for the current chain.
+    pub fn fetch_shards(&self, options: &ShardingOptions) -> Result<Vec<Shard>> {
```

Comment:
I find this API a bit weird. I think a more natural workflow would be:
- fetch the manifest
- from the manifest, select the archives that match the query criteria (ie. chain, shard id, etc.)
- fetch each archive, unpack and return the collection of contracts

It seems to me that the `Repository` struct is redundant. I'd center the API around the `Manifest`, the `ContractArchive`, and of course the `Contract` (as right now).